### DISCUSSION TITLE
passwordless authentication

### DISCUSSION THREAD

**Discussion by Akabeilaa** [community/community#125484] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

When registering Passkey in the website, call the API interface navigator.credentials.create (), and set the publicKey.authenticatorselection.UserVervice to discouraged. It can be successfully registered with virtual USB. But when the parameter is changed to Required, it is displayed: this security password cannot be used. Is it possible that the UserVerification does not support USB login for Required?

**Comment by GylanSalih** [comment]
When using the WebAuthn API to register a passkey, the userVerification parameter within the authenticatorSelection dictates how user verification should be handled during the registration process. The parameter can have the following values:

    required
    preferred
    discouraged

If setting userVerification to required results in an error, it's possible that the authenticator you're using does not support user verification or does not have a method to verify the user (such as biometrics or PIN) that satisfies the required condition. Virtual USB authenticators may not support all features necessary for user verification, particularly if they lack mechanisms like biometrics or PIN entry.

Here's a brief explanation of the different userVerification settings:

    required: The authenticator must perform user verification. If the authenticator cannot verify the user (e.g., through biometrics or PIN), it cannot be used.

    preferred: The authenticator will attempt to perform user verification if it can, but it's not mandatory. If it can't, it will still proceed without it.

    discouraged: User verification should not be performed. This is useful when user verification is handled separately by the web application.

Troubleshooting Steps

    Check Authenticator Capabilities:
        Ensure the authenticator supports user verification. Physical USB keys often support user verification through a PIN or biometric verification, while some virtual or less sophisticated USB authenticators may not.

    Review Authenticator Configuration:
        Verify the settings and capabilities of your virtual USB authenticator. Some software-based virtual USB authenticators may not support user verification as required by the required setting.

    Test with a Physical Security Key:
        If possible, test the registration process with a physical security key known to support user verification. This can help determine if the issue is specific to the virtual USB authenticator.

    Review Error Messages:
        Check the exact error message returned when userVerification is set to required. This may provide more insight into why the registration fails.

Example Code

Here is an example of how you might configure navigator.credentials.create() with userVerification set to required:

javascript

navigator.credentials.create({
    publicKey: {
        challenge: new Uint8Array([/* challenge data from server */]),
        rp: {
            name: "Example Corp"
        },
        user: {
            id: new Uint8Array([/* user ID */]),
            name: "username@example.com",
            displayName: "User Name"
        },
        pubKeyCredParams: [{
            type: "public-key",
            alg: -7
        }],
        authenticatorSelection: {
            authenticatorAttachment: "cross-platform",
            userVerification: "required"
        },
        timeout: 60000,
        attestation: "direct"
    }
}).then(function (newCredentialInfo) {
    // handle successful registration
}).catch(function (err) {
    console.error(err);
    // handle errors
});

Conclusion

If userVerification set to required fails with your virtual USB authenticator, it likely indicates that the authenticator does not support the required user verification methods. Testing with a physical security key that supports user verification can help confirm this. If the issue persists, consulting the documentation or support for the specific authenticator being used would be beneficial.

**Reply by Akabeilaa** [reply]
Thanks,i will try it again
